playbooks/common/openshift-master/restart.yml

```diff
@@ -0,0 +1,163 @@
+---
+- include: ../openshift-cluster/evaluate_groups.yml
+
+- name: Validate configuration for rolling restart
+  hosts: oo_masters_to_config
+  roles:
+  - openshift_facts
+  tasks:
+  - fail:
+      msg: "openshift_rolling_restart_mode must be set to either 'services' or 'system'"
+    when: openshift_rolling_restart_mode is defined and openshift_rolling_restart_mode not in ["services", "system"]
+  - openshift_facts:
+      role: "{{ item.role }}"
+      local_facts: "{{ item.local_facts }}"
+    with_items:
+      - role: common
+        local_facts:
+          rolling_restart_mode: "{{ openshift_rolling_restart_mode | default('services') }}"
+      - role: master
+        local_facts:
+          cluster_method: "{{ openshift_master_cluster_method | default(None) }}"
+
+# Creating a temp file on localhost, we then check each system that will
+# be rebooted to see if that file exists, if so we know we're running
+# ansible on a machine that needs a reboot, and we need to error out.
+- name: Create temp file on localhost
+  hosts: localhost
+  connection: local
+  become: no
+  gather_facts: no
+  tasks:
+  - local_action: command mktemp
+    register: mktemp
+    changed_when: false
+
+- name: Check if temp file exists on any masters
+  hosts: oo_masters_to_config
+  tasks:
+  - stat: path="{{ hostvars.localhost.mktemp.stdout }}"
+    register: exists
+    changed_when: false
+
+- name: Cleanup temp file on localhost
+  hosts: localhost
+  connection: local
+  become: no
+  gather_facts: no
+  tasks:
+  - file: path="{{ hostvars.localhost.mktemp.stdout }}" state=absent
+    changed_when: false
+
+- name: Fail if restarting the system where ansible is running
+  hosts: oo_masters_to_config
+  any_errors_fatal: true
+  tasks:
+  - fail: msg="Cannot run playbook on a host that will be restarted."
+    when: exists.stat.exists
+
+- name: Restart masters (native)
+  hosts: oo_masters_to_config
+  serial: 1
+  tasks:
+  - name: Restart master system
```

Comment:
Okay, I'm starting to think that we should have a task file that we include that does the actual restart, we can conditionally include a task file for restarting the host or one for restarting the services for each type of cluster_method

Then we could do something like this:
```
- name: Restart masters
  hosts: oo_masters_to_config
  serial: 1
  tasks:
  - include: restart_host.yml
    when: (openshift.common.rolling_restart_mode == 'system' and openshift.master.cluster_method == 'native') or (openshift.master.cluster_method == 'pacemaker' and not is_active)

  - include: restart_native_services.yml
    when: openshift.master.cluster_method == 'native'

  - include: restart_pacemaker_services.yml
    when: openshift.common.rolling_restart_mode == 'service' and openshift.master.cluster_method == 'pacemaker' and not is_active
```
post_tasks:
- include: wait_for_native_service.yml
  when: openshift.master.cluster_method == 'native'
- include: wait_for_pacemaker_services.yml
  when: openshift.master.cluster_method == 'pacemaker' and not is_active

We would also need to shift around the pacemaker active logic and still perform the restart on the active host last. 

Think,ing about it a bit more, we should probably just use add_host to create 3 groups and have a playbook that does the above for each...
first group would be all the hosts we want to just roll through (native ha hosts + pacemaker hosts that aren't active) 
- the current host
second group would be pacemaker hosts that are active
third group would consist of the current host.

Then we could push all the restart and validation logic into a playbook that we include 3 times and pass in the list of hosts that comprise the appropriate group.

Man, I'm longing for ansible v2 and being able to loop over a block...